### PR TITLE
[dag] update dag fetcher to construct request inside

### PR DESCRIPTION
**Pull request by zekun000**

### CODE REVIEW

File: consensus/src/dag/dag_store.rs [hunk 1]
```diff
@@ -115,4 +118,9 @@ impl Dag {
             None
         }
     }
+
+    pub fn bitmask(&self) -> Vec<Vec<bool>> {
+        // TODO: extract local bitvec
+        vec![]
```
Comment (ibalajiarun) [inline consensus/src/dag/dag_store.rs:124]:
```suggestion
        todo!()
```

File: consensus/src/dag/dag_fetcher.rs [hunk 2]
```diff
@@ -19,44 +19,51 @@ use tokio::sync::{
     oneshot,
 };
 
-enum FetchCallback {
+pub enum LocalFetchRequest {
```
Comment (ibalajiarun) [inline consensus/src/dag/dag_fetcher.rs:22]: This is more like FetchState than a Request no?

File: consensus/src/dag/dag_store.rs [hunk 3]
```diff
@@ -79,8 +79,11 @@ impl Dag {
         self.nodes_by_digest.contains_key(digest)
     }
 
-    pub fn all_exists<'a>(&self, mut digests: impl Iterator<Item = &'a HashValue>) -> bool {
-        digests.all(|digest| self.nodes_by_digest.contains_key(digest))
+    pub fn all_exists(&self, nodes: &[NodeCertificate]) -> bool {
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/dag/dag_store.rs:82]: I think most of cases we only use this for node.parents(), that's why i changed it

File: consensus/src/dag/dag_fetcher.rs (hunk 2 above)
Comment (zekun000) [reply consensus/src/dag/dag_fetcher.rs:22]: in a sense, it's a request sending from local component like dag_driver

File: consensus/src/dag/dag_fetcher.rs [hunk 4]
```diff
@@ -19,44 +19,51 @@ use tokio::sync::{
...
+
+    pub fn node(&self) -> &Node {
+        match self {
+            LocalFetchRequest::Node(node, _) => node,
+            LocalFetchRequest::CertifiedNode(node, _) => node,
+        }
+    }
 }
 
 struct DagFetcher {
```
[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/dag/dag_fetcher.rs:54]: i feel it's not enough functionality to test yet, it's a very thin wrapper around the dag right now

File: consensus/src/dag/types.rs [hunk 5]
```diff
@@ -386,12 +386,22 @@ impl BroadcastStatus for CertificateAckState {
...
 #[derive(Serialize, Deserialize, Clone, Debug)]
-pub struct FetchRequest {
+pub struct RemoteFetchRequest {
     target: NodeMetadata,
     start_round: Round,
     exists_bitmask: Vec<Vec<bool>>,
 }
 
+impl RemoteFetchRequest {
+    pub fn new(target: NodeMetadata, start_round: Round, exists_bitmask: Vec<Vec<bool>>) -> Self {
```
Comment (sasha8) [inline consensus/src/dag/types.rs:396]: Why "exists"?

**Review by sasha8 — APPROVED** [review]

**Review by ibalajiarun — APPROVED** [review]

**Pull request merged**